### PR TITLE
sort matrix according to issue description

### PR DESCRIPTION
**Pull request by Flow86**
see #1581

### CODE REVIEW

File: .github/workflows/unit-tests.yml [hunk 1]
```diff
@@ -121,7 +121,6 @@ jobs:
       - name: Install dependencies
         if: "!startsWith(runner.os, 'macos')"
         run: |
-          # Minimum requirements: CMake 3.9, Boost 1.69
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:124]: IMO this is kind of important information. We currently don't have a test for CMake 3.9 at all anymore. Hence I'd rather make that the default in the global env. Same for Boost: 1.69 is the minimum so why use another version as the default?

File: .github/workflows/unit-tests.yml [hunk 2]
```diff
@@ -23,21 +23,21 @@ jobs:
       matrix:
         include:
           # MacOSX
-          - { compiler: clang,    os: macos-12,     buildType: Debug, cxxStandard: 14 }
+          - { compiler: clang,    os: macos-12,     buildType: Debug }
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:26]: IMO it makes sense to fix/set the Boost version here as it is not arbitrary but depends on the choice of macos version. See also the debug output of https://github.com/Return-To-The-Roots/s25client/pull/1580/files#diff-cd0effce06f425599e96952f4d5f684641a683f46d9909e7c74a9f8b2f03ccd3R113-R114 which may help when something changes. So I'd include that (debug output and explicit version here)

File: .github/workflows/unit-tests.yml [hunk 3]
```diff
@@ -23,21 +23,21 @@ jobs:
...
           # Linux GCC
           # GCC 9 is known to show a few warnings that GCC 10 has "fixed", make sure this doesn't happen for us
-          - { compiler: gcc-9,    os: ubuntu-20.04, buildType: Debug }
-          - { compiler: gcc-10,   os: ubuntu-20.04, buildType: Debug, cxxStandard: 14, coverage: true }
-          - { compiler: gcc-10,   os: ubuntu-20.04, buildType: Debug, cxxStandard: 17 }
-          - { compiler: gcc-10,   os: ubuntu-20.04, buildType: Release, cxxStandard: 14 }
+          - { compiler: gcc-9,    os: ubuntu-20.04, buildType: Debug, boostVersion: 1.69.0 }
+          - { compiler: gcc-10,   os: ubuntu-20.04, buildType: Debug, boostVersion: 1.69.0, coverage: true }
+          - { compiler: gcc-10,   os: ubuntu-20.04, buildType: Release }
+          - { compiler: gcc-11,   os: ubuntu-22.04, buildType: Debug, boostVersion: 1.74.0 }
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:32]: I think it makes sense noting that GCC < 11 is C++14 and GCC >= 11 is C++17 at least as a comment so it is clear we test both. Similar for clang. You already gathered that info in the issue but that will be lost/hard to find when this is merged. Maybe even make it explicit by using `cxxStandard: 14` for the gcc-10 Release build and `cxxStandard: 17` for the gcc-12 release build.

File: .github/workflows/unit-tests.yml [hunk 4]
```diff
@@ -23,21 +23,21 @@ jobs:
...
           # Linux clang
-          - { compiler: clang-11, os: ubuntu-20.04, buildType: Debug, cxxStandard: 14, externalSanitizer: true }
-          - { compiler: clang-11, os: ubuntu-20.04, buildType: Debug, cxxStandard: 17 }
-          - { compiler: clang-11, os: ubuntu-20.04, buildType: Release, cxxStandard: 14 }
+          - { compiler: clang-10, os: ubuntu-20.04, buildType: Debug, boostVersion: 1.69.0, externalSanitizer: true }
+          - { compiler: clang-10, os: ubuntu-20.04, buildType: Release }
+          - { compiler: clang-14, os: ubuntu-22.04, buildType: Debug, boostVersion: 1.74.0 }
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:36]: I'd add a comment here why Boost 1.74 is used

File: .github/workflows/unit-tests.yml [hunk 5]
```diff
@@ -23,21 +23,21 @@ jobs:
...
           # Linux Latest GCC
-          - { compiler: gcc-12,   os: ubuntu-22.04, buildType: Debug, cxxStandard: 14, coverage: true }
+          - { compiler: gcc-12,   os: ubuntu-22.04, buildType: Debug, boostVersion: 1.82.0 }
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:38]: Same here, e.g.:
```suggestion
          - { compiler: gcc-12,   os: ubuntu-22.04, buildType: Debug, boostVersion: 1.82.0 } # GCC 12 detects an issue fixed in Boost 1.82
```

File: .github/workflows/unit-tests.yml [hunk 6]
```diff
@@ -23,21 +23,21 @@ jobs:
...
           # Linux Latest clang
-          - { compiler: clang-16, os: ubuntu-20.04, buildType: Debug, cxxStandard: 14, externalSanitizer: true }
+          - { compiler: clang-16, os: ubuntu-20.04, buildType: Debug }
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:40]: I'd keep the sanitizer enabled: So we have it for the oldest and newest supported clang

**Review by Flamefire — CHANGES_REQUESTED** [review]
Looks good to me, just missing a bit of information in the file WHY a specific version was chosen, especially the "global" Boost/CMake version look very random.

See the inline comments.

File: .github/workflows/unit-tests.yml (hunk 1 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:124]: thats exactly one point, we should shift the minimum up to get rid of all the workarounds. 

Stock cmake from debian bullseye is 3.18.4 - we should probably make this the new minimum

File: .github/workflows/unit-tests.yml (hunk 6 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:40]: oh yes, that was by accident

File: .github/workflows/unit-tests.yml (hunk 3 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:32]: I wrote it down now in the comments

File: .github/workflows/unit-tests.yml (hunk 2 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:26]: I wrote it down now in the comments

File: .github/workflows/unit-tests.yml (hunk 4 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:36]: I wrote it down now in the comments

File: .github/workflows/unit-tests.yml (hunk 5 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:38]: I wrote it down now in the comments

File: .github/workflows/unit-tests.yml [hunk 7]
```diff
@@ -22,39 +23,76 @@ jobs:
...
+          - { compiler: clang-10, os: ubuntu-20.04, type: Debug, cmake: 3.16.3, boost: 1.69.0, externalSanitizer: true }
+          #
+          #   Default compiler for Ubuntu 20.04
+          #   Use (compiler) default C++ standard (14)
+          #   Use default cmake
+          #   Use default boost
+          - { compiler: gcc-10,   os: ubuntu-20.04, type: Debug, coverage: true }
+          #
+          #   Default compilers for Ubuntu 22.04
+          #   Use C++ standard (17, default for gcc)
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:49]: C&P made this hard to understand, maybe: 
```suggestion
          #   Use C++17 standard (default for gcc-11)
```

File: .github/workflows/unit-tests.yml [hunk 8]
```diff
@@ -22,39 +23,76 @@ jobs:
...
 
     steps:
       - run: echo "DEPS_DIR=${{runner.temp}}/.cache" >> $GITHUB_ENV
-      - run: echo "CMAKE_VERSION=${{matrix.cmakeVersion}}" >> $GITHUB_ENV
-        if: matrix.cmakeVersion
-      - run: echo "BOOST_VERSION=${{matrix.boostVersion}}" >> $GITHUB_ENV
-        if: matrix.boostVersion
-      - run: echo "ADDITIONAL_CMAKE_FLAGS=${ADDITIONAL_CMAKE_FLAGS} -DRTTR_EXTERNAL_BUILD_TESTING=ON -DRTTR_ENABLE_SANITIZERS=ON" >> $GITHUB_ENV
+
+      - name: "Set BUILD_TYPE"
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:68]: IMO giving this names is overly verbose but fine for me. At least get rid of the quotes (here and below) as in `name: Set C++ standard`:
```suggestion
      - name: Set build type (Release/Debug/...)
```

File: .github/workflows/unit-tests.yml [hunk 9]
```diff
@@ -22,39 +23,76 @@ jobs:
...
+
+      - name: Enable coverage collection
+        if: matrix.coverage
+        run: echo "ADDITIONAL_CMAKE_FLAGS=${ADDITIONAL_CMAKE_FLAGS} -DRTTR_ENABLE_COVERAGE=ON" >> $GITHUB_ENV
+
+      - name: "Enable external sanitizer"
         if: matrix.externalSanitizer
+        run: echo "ADDITIONAL_CMAKE_FLAGS=${ADDITIONAL_CMAKE_FLAGS} -DRTTR_EXTERNAL_BUILD_TESTING=ON -DRTTR_ENABLE_SANITIZERS=ON" >> $GITHUB_ENV
 
+      # Coverage collection requires access to history to find merge-base
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:91]: Why move this above the part that does **not** apply to coverage? Then I'd swap the 2 so this looks more logical and maybe add a comment like  `fetch-depth: 0  # Faster` or so while at it

File: .github/workflows/unit-tests.yml [hunk 10]
```diff
@@ -109,52 +149,48 @@ jobs:
...
-
-      - name: Set C++ standard
-        if: matrix.cxxStandard
-        run: echo "ADDITIONAL_CMAKE_FLAGS=${ADDITIONAL_CMAKE_FLAGS} -DCMAKE_CXX_STANDARD=${{matrix.cxxStandard}}" >> $GITHUB_ENV
-
       - name: Setup environment
-        run: |
-          echo "BUILD_TYPE=${{matrix.buildType}}" >> $GITHUB_ENV
-          echo "TRAVIS_BUILD_DIR=$GITHUB_WORKSPACE" >> $GITHUB_ENV
+        run: echo "TRAVIS_BUILD_DIR=$GITHUB_WORKSPACE" >> $GITHUB_ENV
```
Comment (Flamefire) [inline .github/workflows/unit-tests.yml:180]: As we'll never use travis again, maybe rather change that in the script(s) and get rid of this step which is now only for `$TRAVIS_BUILD_DIR` and rename `tools/ci/travisBuild.sh` to `githubBuild.sh` or just `build.sh`?

This is an ugly clutch anyway the build script should have been able to handle either variable.

File: libs/s25main/Debug.cpp [hunk 11]
```diff
@@ -62,16 +62,16 @@ bool captureBacktrace(std::vector<void*>& stacktrace, LPCONTEXT ctx = nullptr) n
...
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wcast-function-type"
-    // error: cast between incompatible function types from
-    // 'FARPROC' {aka 'int (__attribute__((stdcall)) *)()'}
-    // to
-    // 'RtlCaptureContextType' {aka 'void (__attribute__((stdcall)) *)(CONTEXT*)'}
-    // [-Werror=cast-function-type]
-    // and so on
-#endif
+#            if __GNUC__ >= 9
```
Comment (Flamefire) [inline libs/s25main/Debug.cpp:65]: I do agree this looks odd/ugly here but serves the point to highlight that it is in multiple conditions. And e.g. https://github.com/Return-To-The-Roots/s25client/pull/1582/files#diff-e93d2e7c03a97a61287ef951276877e6d0cacac56ce2e0daecc3c935bae907eaR24-R35 looks much better due to this.

I'm wondering if we couldn't use Boost for this (backtrace collection) anyway or at least add a wrapper around `GetProcAddress` to do the casting with the warning disabled or use Boost.DLL which IIRC can do that too.

But that's another PR...

File: .github/workflows/unit-tests.yml (hunk 1 above)
Comment (Flamefire) [reply .github/workflows/unit-tests.yml:124]: Ubuntu 20 having CMake 3.16 is a reasonable argument to having that as the minimum as IIRC Ubuntu 18 is EOL. Maybe mention that somewhere but we shouldn't increase the required version if there isn't a benefit to not offend/inconvenience potential contributors.

**Review by Flamefire — COMMENTED** [review]
This has gotten a bit verbose increasing vertical spacing a lot. But whatever. A few remarks inline

File: .github/workflows/unit-tests.yml (hunk 9 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:91]: so that the comment is above the block

File: .github/workflows/unit-tests.yml (hunk 8 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:68]: oh there were some with quotes and some without. Ive removed it

File: .github/workflows/unit-tests.yml (hunk 7 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:49]: good point

File: .github/workflows/unit-tests.yml (hunk 10 above)
Comment (Flow86) [reply .github/workflows/unit-tests.yml:180]: I don't know if the script is used at another place, so I'll leave it like it is for now

**Pull request merged**